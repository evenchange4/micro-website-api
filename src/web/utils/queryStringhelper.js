/* eslint no-underscore-dangle: 0 */

import * as R from 'ramda';
import uuidv4 from 'uuid/v4';
import qs from 'query-string';

const queryStringhelper = {
  parse: R.pipe(qs.parse, parsed => ({
    url: R.defaultTo('')(parsed.url),
    selector: R.defaultTo('')(parsed.selector),
    cache: R.cond([
      [R.equals('true'), R.always(true)],
      [R.equals('false'), R.always(false)],
      [R.T, R.always(true)],
    ])(parsed.cache),
    format: R.cond([
      [R.contains(R.__, ['raw', 'json']), R.identity],
      [R.T, R.always('raw')],
    ])(parsed.format),
    actions: R.pipe(
      R.defaultTo([]),
      R.cond([[R.is(String), R.of], [R.T, R.identity]]),
      R.map(R.split(',')),
      R.map(R.zipObj(['click', 'wait'])),
      R.map(e => R.assoc('key', uuidv4(), e)),
    )(parsed.actions),
  })),
  stringify: R.pipe(
    R.evolve({
      url: encodeURIComponent,
      selector: encodeURIComponent,
      cache: encodeURIComponent,
      format: encodeURIComponent,
      actions: R.map(
        R.pipe(
          R.omit(['key']),
          R.values,
          R.map(encodeURIComponent),
          R.join(','),
        ),
      ),
    }),
    v => qs.stringify(v, { encode: false }),
  ),
};

export default queryStringhelper;
