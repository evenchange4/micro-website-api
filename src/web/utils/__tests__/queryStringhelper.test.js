import queryStringhelper from '../queryStringhelper';

jest.mock('uuid/v4', () => () => 'mockUuidv4');

it('should return correct values with queryStringhelper.parse', () => {
  [
    'url=',
    'url=url',
    'url=url&selector=',
    'url=url&selector=selector',

    'cache=',
    'cache=true',
    'cache=false',
    'cache=other',

    'format=',
    'format=raw',
    'format=json',
    'format=other',

    'actions=',
    'actions=click,',
    'actions=,wait',
    'actions=click,wait',
    'actions=click,wait&actions=',
    'actions=click,wait&actions=click,',
    'actions=click,wait&actions=click,wait',
  ].forEach(e => {
    const values = queryStringhelper.parse(e);
    expect(values).toMatchSnapshot(e);
    expect(
      queryStringhelper.parse(queryStringhelper.stringify(values)),
    ).toEqual(values);
  });
});
