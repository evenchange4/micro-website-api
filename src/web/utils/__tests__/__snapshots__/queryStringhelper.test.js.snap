// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions= 1`] = `
Object {
  "actions": Array [
    Object {
      "click": "",
      "key": "mockUuidv4",
    },
  ],
  "cache": true,
  "format": "raw",
  "selector": "",
  "url": "",
}
`;

exports[`actions=,wait 1`] = `
Object {
  "actions": Array [
    Object {
      "click": "",
      "key": "mockUuidv4",
      "wait": "wait",
    },
  ],
  "cache": true,
  "format": "raw",
  "selector": "",
  "url": "",
}
`;

exports[`actions=click, 1`] = `
Object {
  "actions": Array [
    Object {
      "click": "click",
      "key": "mockUuidv4",
      "wait": "",
    },
  ],
  "cache": true,
  "format": "raw",
  "selector": "",
  "url": "",
}
`;

exports[`actions=click,wait 1`] = `
Object {
  "actions": Array [
    Object {
      "click": "click",
      "key": "mockUuidv4",
      "wait": "wait",
    },
  ],
  "cache": true,
  "format": "raw",
  "selector": "",
  "url": "",
}
`;

exports[`actions=click,wait&actions= 1`] = `
Object {
  "actions": Array [
    Object {
      "click": "click",
      "key": "mockUuidv4",
      "wait": "wait",
    },
    Object {
      "click": "",
      "key": "mockUuidv4",
    },
  ],
  "cache": true,
  "format": "raw",
  "selector": "",
  "url": "",
}
`;

exports[`actions=click,wait&actions=click, 1`] = `
Object {
  "actions": Array [
    Object {
      "click": "click",
      "key": "mockUuidv4",
      "wait": "wait",
    },
    Object {
      "click": "click",
      "key": "mockUuidv4",
      "wait": "",
    },
  ],
  "cache": true,
  "format": "raw",
  "selector": "",
  "url": "",
}
`;

exports[`actions=click,wait&actions=click,wait 1`] = `
Object {
  "actions": Array [
    Object {
      "click": "click",
      "key": "mockUuidv4",
      "wait": "wait",
    },
    Object {
      "click": "click",
      "key": "mockUuidv4",
      "wait": "wait",
    },
  ],
  "cache": true,
  "format": "raw",
  "selector": "",
  "url": "",
}
`;

exports[`cache= 1`] = `
Object {
  "actions": Array [],
  "cache": true,
  "format": "raw",
  "selector": "",
  "url": "",
}
`;

exports[`cache=false 1`] = `
Object {
  "actions": Array [],
  "cache": false,
  "format": "raw",
  "selector": "",
  "url": "",
}
`;

exports[`cache=other 1`] = `
Object {
  "actions": Array [],
  "cache": true,
  "format": "raw",
  "selector": "",
  "url": "",
}
`;

exports[`cache=true 1`] = `
Object {
  "actions": Array [],
  "cache": true,
  "format": "raw",
  "selector": "",
  "url": "",
}
`;

exports[`format= 1`] = `
Object {
  "actions": Array [],
  "cache": true,
  "format": "raw",
  "selector": "",
  "url": "",
}
`;

exports[`format=json 1`] = `
Object {
  "actions": Array [],
  "cache": true,
  "format": "json",
  "selector": "",
  "url": "",
}
`;

exports[`format=other 1`] = `
Object {
  "actions": Array [],
  "cache": true,
  "format": "raw",
  "selector": "",
  "url": "",
}
`;

exports[`format=raw 1`] = `
Object {
  "actions": Array [],
  "cache": true,
  "format": "raw",
  "selector": "",
  "url": "",
}
`;

exports[`url= 1`] = `
Object {
  "actions": Array [],
  "cache": true,
  "format": "raw",
  "selector": "",
  "url": "",
}
`;

exports[`url=url 1`] = `
Object {
  "actions": Array [],
  "cache": true,
  "format": "raw",
  "selector": "",
  "url": "url",
}
`;

exports[`url=url&selector= 1`] = `
Object {
  "actions": Array [],
  "cache": true,
  "format": "raw",
  "selector": "",
  "url": "url",
}
`;

exports[`url=url&selector=selector 1`] = `
Object {
  "actions": Array [],
  "cache": true,
  "format": "raw",
  "selector": "selector",
  "url": "url",
}
`;
